src/createClient.js

```diff
@@ -1,44 +1,58 @@
-import makeUUID from 'uuid/v4';
-import { connect, publishStep, stepsAligned, stepSucceeded, stepFailed } from './messages';
+// @flow
 
+import uuid from 'uuid';
 
-export default function (clientId, clientDescription = '') {
+
+export default function (clientId: string, clientDescription: string = ''): ChorusClient {
 	if (!clientId) {
 		throw new Error('Please provide `clientId`');
 	}
 
-	let _socket = null;
-	const _callbacks = {};
+	let _socket: WebSocket;
+	const _callbacks: {[stepId: string]: (...args: Array<string>) => number | string | void} = {};
 
-	function _sendMessage(message) {
+	function _sendMessage(message: OutgoingMessage): void {
+		if (!_socket) { return; }
 		_socket.send(
 			JSON.stringify(message),
 		);
 	}
 
-	function _onMessage(event) {
-		const message = JSON.parse(event.data);
+	function _onMessage(event: Event) {
+		if (typeof event.data !== 'string') {
+			throw new Error('Expecting websocket message data to be of type string');
+		}
+		const incomingMessage: ExecuteStepMessage = JSON.parse(event.data);
+
+		if (incomingMessage.type === 'EXECUTE_STEP') {
+			const callback = _callbacks[incomingMessage.stepId];
+			const { stepId, executionId, contextVariables } = incomingMessage;
 
-		if (message.type === 'EXECUTE_STEP') {
-			const callback = _callbacks[message.stepId];
-			if (callback) {
-				const { stepId, executionId, contextVariables } = message;
-				try {
-					const result = callback(...message.arguments);
-					_sendMessage(
-						stepSucceeded(clientId, stepId, executionId, result, contextVariables),
-					);
-				} catch (e) {
-					_sendMessage(
-						stepFailed(clientId, stepId, executionId, e.message),
-					);
-				}
+			try {
+				const result: number | string | void = callback(...incomingMessage.arguments);
+				const message: StepSucceededMessage = {
```

Comment:
Flow doesn't enforce anything.

However, what I had before was an extra file with what I called **message creators**. Basically simple functions that would take arguments one by one and return an object constructed correctly.

I felt that now that the messages are typed, it was redundant to have an extra step as the functions add an extra level of potential errors. In particular even if I typed the message creator function parameters, one could still easily put them in the wrong order for example.

Whereas with the typed message as we have them now, it's guaranteed that the message is correct otherwise it wouldn't typecheck properly.